File: src/algorithm_java/leetcode/Problem105.java
package algorithm.leetcode;

public class Problem105 {

    public static TreeNode buildTree(int[] preorder, int[] inorder) {
        return build(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1);
    }

    public static TreeNode build(int[] preorder, int pLeft, int pRight, int[] inorder, int inLeft, int inRight) {
        if (pLeft > pRight || inLeft > inRight)
            return null;
        TreeNode root = new TreeNode(preorder[pLeft]);
        int rootIdxInInorder = inLeft;
        for (; rootIdxInInorder <= inRight; rootIdxInInorder++) {
            if (inorder[rootIdxInInorder] == preorder[pLeft])
                break;
        }
        root.left = build(preorder, pLeft + 1, pLeft + rootIdxInInorder - inLeft,
                inorder, inLeft, rootIdxInInorder - 1);
        root.right = build(preorder, pLeft + rootIdxInInorder - inLeft + 1, pRight,
                inorder, rootIdxInInorder + 1, inRight);
        return root;
    }

    public static void main(String[] args) {
        int[] preorder = new int[]{3,9,20,15,7};
        int[] inorder = new int[]{9,3,15,20,7};
        TreeNode root = buildTree(preorder, inorder);
        System.out.println();
    }
}

File: src/algorithm_cpp/problem105.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if (preorder.size() == 0)
            return nullptr;
        TreeNode* root = new TreeNode(preorder[0]);
        vector<int> leftpre, leftin, rightpre, rightin;
        bool flag = false;
        for (int i = 0; i < inorder.size(); i++) {
            if (inorder[i] == preorder[0]) {
                flag = true;
                continue;
            }
            if (flag == false) leftin.push_back(inorder[i]);
            if (flag == true) rightin.push_back(inorder[i]);
        }
        for (int i = 0; i < leftin.size(); i++) {
            leftpre.push_back(preorder[1 + i]);
        }
        for (int i = 0; i < rightin.size(); i++) {
            rightpre.push_back(preorder[1 + leftin.size() + i]);
        }
        root->left = buildTree(leftpre, leftin);
        root->right = buildTree(rightpre, rightin);
        return root;
    }
};
